contracts/V4Router.sol

```diff
@@ -0,0 +1,189 @@
+        _swapExactPrivate(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            msgSender,
+            true,
+            true,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params, address msgSender) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+
+            for (uint256 i = 0; i < pathLength; i++) {
+                (PoolKey memory poolKey, bool zeroForOne) = _getPoolAndSwapDirection(params.path[i], params.currencyIn);
+                amountOut = uint128(
+                    -_swapExactPrivate(
+                        poolKey,
+                        zeroForOne,
+                        int256(int128(params.amountIn)),
+                        0,
+                        msgSender,
+                        i == 0,
+                        i == pathLength - 1,
+                        params.path[i].hookData
+                    )
+                );
+
+                params.amountIn = amountOut;
+                params.currencyIn = params.path[i].intermediateCurrency;
+            }
+
+            if (amountOut < params.amountOutMinimum) revert TooLittleReceived();
+        }
+    }
+
+    function _swapExactOutputSingle(IV4Router.ExactOutputSingleParams memory params, address msgSender) private {
+        _swapExactPrivate(
+            params.poolKey,
+            params.zeroForOne,
+            -int256(int128(params.amountOut)),
+            params.sqrtPriceLimitX96,
+            msgSender,
+            true,
+            true,
+            params.hookData
+        );
+    }
+
+    function _swapExactOutput(IV4Router.ExactOutputParams memory params, address msgSender) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountIn;
+
+            for (uint256 i = pathLength; i > 0; i--) {
+                (PoolKey memory poolKey, bool oneForZero) =
+                    _getPoolAndSwapDirection(params.path[i - 1], params.currencyOut);
+                amountIn = uint128(
+                    _swapExactPrivate(
+                        poolKey,
+                        !oneForZero,
+                        -int256(int128(params.amountOut)),
+                        0,
+                        msgSender,
+                        i == 1,
+                        i == pathLength,
+                        params.path[i - 1].hookData
+                    )
+                );
+
+                params.amountOut = amountIn;
+                params.currencyOut = params.path[i - 1].intermediateCurrency;
+            }
+            if (amountIn > params.amountInMaximum) revert TooMuchRequested();
+        }
+    }
+
+    function _swapExactPrivate(
```

Comment:
can we name this like `_swap`? It differentiates it from the other entrypoints and is a bit more concise